[jsk_fetch_startup] Add save_battery node to check if shutdown is needed

nakane11 opened the pull request:
This PR prevents shutdown if the robot is manually docked after kitchen demo has failed.

nakane11 committed:
Add node to check if shutdown is needed

nakane11 committed:
Publish /save_battery instead of /shutdown

nakane11 committed:
Rename topic to shutdown_unchecked and node to uncharged_shutdown

nakane11 commented:
I checked `/shutdown` topic is only published when `is_charging` is not true.

708yamaguchi commented on jsk_fetch_robot/jsk_fetch_startup/scripts/uncharged_shutdown.py:
```diff
@@ -0,0 +1,27 @@
+
+    def __init__(self):
+        rospy.loginfo('Start uncharged_shutdown node.')
+        self.timeout = 1
```
`battery_state` is published at 1 hz, so timeout should be longer than 1[s]. For example, 2[s].
```
fetch@fetch15:~$ rostopic hz /battery_state
subscribed to [/battery_state]
no new messages
average rate: 1.000
	min: 1.000s max: 1.000s std dev: 0.00000s window: 2
average rate: 1.000
	min: 1.000s max: 1.000s std dev: 0.00032s window: 3
```

708yamaguchi replied:
I think it would be better
- to set `timeout` rosparam.
or
- to use fixed timeout value. (e.g. `rospy.wait_for_message('/battery_state', BatteryState, timeout=2)`)

708yamaguchi commented on jsk_fetch_robot/jsk_fetch_startup/scripts/uncharged_shutdown.py:
```diff
@@ -0,0 +1,27 @@
+                return
+        except:
+            pass
+        self.shutdown_pub.publish(Empty())
```
I am happy if fetch speaks and outputs loginfo before shutdown.
For example, 
- バッテリ保護のためシャットダウンします。loginfo("Shutdown to save battery")
- 充電されているのでシャットダウンはしません。loginfo("I received shutdown_uncheck, but dont shutdown because I am charged")

708yamaguchi commented on jsk_fetch_robot/jsk_fetch_startup/scripts/uncharged_shutdown.py:
```diff
@@ -0,0 +1,27 @@
+        self.shutdown_pub = rospy.Publisher('/shutdown', Empty, queue_size=1)
+
+    def uncharged_shutdown(self, msg):
+        try:
```
Why do you use `try ... except`?
At least you can specify exception type.
FYI
https://www.flake8rules.com/rules/E722.html

708yamaguchi reviewed (commented):
Thank you!
Nice feature.

I will check this node with real fetch robot.

nakane11 committed:
[jsk_fetch_startup] Fix timeout value and specify exception

nakane11 committed:
[jsk_fetch_startup] Add speak and loginfo

nakane11 commented:
@708yamaguchi 
Thank you for your review.
I fixed timeout to 2Hz and specified exception to catch when timeout is exceeded. 248045f  
Added loginfo and speak japanese by 8544dd3

iory commented on jsk_fetch_robot/jsk_fetch_startup/scripts/uncharged_shutdown.py:
```diff
@@ -0,0 +1,53 @@
+            pass
+        rospy.loginfo("I'm going to shut down to protect the battery.")
+        self.speak(self.client_jp,
+                   'バッテリ保護のためシャットダウンします。',
```
How about shutdown.py taking this role?

nakane11 replied:
shutdown.py is placed in jsk_robot_startup, but  I don't know if other robots also publish `/battery_state` or `/robot_state`.

708yamaguchi replied:
> this role

I think @iory is talking about rospy.loginfo and speak role.

708yamaguchi replied:
And I think it is reasonable to move these roles to shutdown.py

iory commented:
既存のテストされているノードを組み合わせて同様の機能が実現できるので、
https://github.com/knorth55/jsk_robot/pull/318　を用いるのが良いかと思います。

nakane11 commented:
ありがとうございます。
シャットダウンの時に話すのも必要ないでしょうか？https://github.com/nakane11/jsk_robot/commit/c904661eaaf3248010af2daab271e6d935230134

iory commented:
あります！
他のPRにして送りましょう！
